pkg/processor/puller.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package processor
+
+import (
+	stdContext "context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/ticdc/cdc/model"
+	"github.com/pingcap/ticdc/cdc/puller"
+	"github.com/pingcap/ticdc/pkg/context"
+	"github.com/pingcap/ticdc/pkg/pipeline"
+	"github.com/pingcap/ticdc/pkg/regionspan"
+	"github.com/pingcap/ticdc/pkg/security"
+	"github.com/pingcap/ticdc/pkg/util"
+	tidbkv "github.com/pingcap/tidb/kv"
+	"golang.org/x/sync/errgroup"
+)
+
+type pullerNode struct {
+	credential *security.Credential
+	kvStorage  tidbkv.Storage
+	limitter   *puller.BlurResourceLimitter
+	tableName  string // quoted schema and table, used in metircs only
+
+	tableID     model.TableID
+	replicaInfo *model.TableReplicaInfo
+	cancel      stdContext.CancelFunc
+	wg          errgroup.Group
+}
+
+func newPullerNode(
+	credential *security.Credential,
+	kvStorage tidbkv.Storage,
+	limitter *puller.BlurResourceLimitter,
+	tableID model.TableID, replicaInfo *model.TableReplicaInfo, tableName string) pipeline.Node {
+	return &pullerNode{
+		credential:  credential,
+		kvStorage:   kvStorage,
+		limitter:    limitter,
+		tableID:     tableID,
+		replicaInfo: replicaInfo,
+		tableName:   tableName,
+	}
+}
+
+func (n *pullerNode) tableSpan(ctx context.Context) []regionspan.Span {
+	// start table puller
+	enableOldValue := ctx.Vars().Config.EnableOldValue
+	spans := make([]regionspan.Span, 0, 4)
+	spans = append(spans, regionspan.GetTableSpan(n.tableID, enableOldValue))
+
+	if ctx.Vars().Config.Cyclic.IsEnabled() && n.replicaInfo.MarkTableID != 0 {
+		spans = append(spans, regionspan.GetTableSpan(n.replicaInfo.MarkTableID, enableOldValue))
+	}
+	return spans
+}
+
+func (n *pullerNode) Init(ctx pipeline.NodeContext) error {
+	enableOldValue := ctx.Vars().Config.EnableOldValue
+	ctxC, cancel := stdContext.WithCancel(ctx.StdContext())
+	ctxC = util.PutTableInfoInCtx(ctxC, n.tableID, n.tableName)
+	plr := puller.NewPuller(ctxC, ctx.Vars().PDClient, n.credential, n.kvStorage,
+		n.replicaInfo.StartTs, n.tableSpan(ctx), n.limitter, enableOldValue)
+	n.wg.Go(func() error {
+		ctx.Throw(errors.Trace(plr.Run(ctxC)))
```

Comment:
Would you please add some comments about when shoud we use pipeline `NodeContext`, and when using standard context is enough?

Reply:
`NodeContext` is created by `pipeline`. If we want to receive a message from the previous node or send a message to the next node, we should use `NodeContext`.
As for `StdContext`, some older functions do not support `cdc/pkg/context.Context`, so we just use `StdContext`